src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,330 @@
+                            documentSymbols.Add(item);
+                        }
+                    }
+
+                    symbolTree.ItemsSource = documentSymbols;
+                }
+            }
+        }
+
+        private void Sort(SortOption sortOption, FunctionId functionId)
+        {
+            if (SymbolTreeInitialized)
+            {
+                Logger.Log(functionId);
+                // If there is an active search, only sort the filtered nodes
+                if (!(searchBox.Text == ServicesVSResources.Search_Document_Outline || string.IsNullOrWhiteSpace(searchBox.Text)))
+                {
+                    symbolTree.ItemsSource = DocumentOutlineHelper.Sort(
+                        new List<DocumentSymbolViewModel>((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource), sortOption);
+                }
+                else
+                {
+                    var sortedDocumentSymbolModels = DocumentOutlineHelper.Sort(SymbolsTreeItemsSource, sortOption);
+                    SymbolsTreeItemsSource = sortedDocumentSymbolModels;
+                    symbolTree.ItemsSource = sortedDocumentSymbolModels;
+                }
+            }
+        }
+
+        private void SortByName(object sender, EventArgs e)
+        {
+            Sort(SortOption.Name, FunctionId.DocumentOutline_SortByName);
+        }
+
+        private void SortByOrder(object sender, EventArgs e)
+        {
+            Sort(SortOption.Order, FunctionId.DocumentOutline_SortByOrder);
+        }
+
+        private void SortByType(object sender, EventArgs e)
+        {
+            Sort(SortOption.Type, FunctionId.DocumentOutline_SortByType);
+        }
+
+        // When symbol node clicked, selects corresponding code
+        private void JumpToContent(object sender, EventArgs e)
+        {
+            RoslynDebug.AssertNotNull(Snapshot);
+            if (sender is StackPanel panel && panel.DataContext is DocumentSymbolViewModel symbol)
+            {
+                if (symbol.StartLine >= 0 && symbol.StartLine < Snapshot.LineCount)
+                {
+                    var position = Snapshot.GetLineFromLineNumber(symbol.StartLine).Start.Position;
+                    if (position >= 0 && position <= Snapshot.Length)
+                    {
+                        TextView.Caret.PositionChanged -= FollowCursor;
+                        var point = new SnapshotPoint(Snapshot, position);
+                        var snapshotSpan = new SnapshotSpan(point, point);
+                        TextView.SetSelection(snapshotSpan);
+                        TextView.ViewScroller.EnsureSpanVisible(snapshotSpan);
+                        TextView.Caret.PositionChanged += FollowCursor;
+                    }
+                }
+            }
+        }
+
+        private void FollowCursor(object sender, EventArgs e)
+        {
+            if (Snapshot is not null && TextView is not null && SymbolsTreeItemsSource is not null)
```

Comment:
this appears to happen entirely on the UI thread.  this may be expensive.  consider moving off of the ui thread.